executor/set_test.go

```diff
@@ -736,6 +736,31 @@ func TestSetVar(t *testing.T) {
 	tk.MustQuery("select @@tidb_cost_model_version").Check(testkit.Rows("2"))
 }
 
+func TestGetSetNoopVars(t *testing.T) {
+	store, clean := testkit.CreateMockStore(t)
+	defer clean()
+	tk := testkit.NewTestKit(t, store)
+
+	// By default you can get/set noop sysvars without issue.
+	tk.MustQuery("SELECT @@query_cache_type").Check(testkit.Rows("OFF"))
+	tk.MustQuery("SHOW VARIABLES LIKE 'query_cache_type'").Check(testkit.Rows("query_cache_type OFF"))
+	tk.MustExec("SET query_cache_type=2")
+	tk.MustQuery("SELECT @@query_cache_type").Check(testkit.Rows("DEMAND"))
+	// When tidb_enable_noop_variables is OFF, you can GET in @@ context
+	// and always SET. But you can't see in SHOW VARIABLES.
+	// Warnings are also returned.
+	tk.MustExec("SET GLOBAL tidb_enable_noop_variables = OFF")
+	defer tk.MustExec("SET GLOBAL tidb_enable_noop_variables = ON")
```

Comment:
FYI, I am not insisted on any change. We don't need to defer anymore, since `store, clean := testkit.CreateMockStore(t)` always creates a brand new separate store/testkit.

Reply:
Good to know! I think I'll leave it here because cleanup is a good practice.